File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int N = nums.length;
        
        int[] dp = new int[N];
        Arrays.fill(dp, 1);

        for (int i = 1; i < N; ++i) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                     dp[i] = Math.max(dp[i], 1 + dp[j]);
                }
            }
        }

        int longest = 0;
        for (int current : dp) {
            longest = Math.max(longest, current);
        }

        return longest;
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
"""



    10,9,2,3,1,5,8

lis =

"""
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        lis = [nums[0]]
        for num in nums[1::]:
            i = bisect_left(lis, num)
            if i == len(lis):
                lis.append(num)
            else:
                lis[i] = num
        return len(lis)
        